### PR TITLE
create a frame around render-to-texture tiles to avoid stiching on different zoomlevels

### PR DESCRIPTION
**Pull request by prozessor13**
This code fixes https://github.com/maplibre/maplibre-gl-js/issues/1602. Creating the frame (e.g. triangle-mesh) to dead-sea around the terrain-tile is not the smartest code, but it works. I dont think that this code has any impact on performance.

@birkskyum it would be great if you can make a render-test for this issue.

### CODE REVIEW

**Comment by HarelM** [comment]
This code is on top of the pool code, right?
To avoid reviewing the same code twice I suggest to target this PR's branch to the branch if the added pool, this way the files that are presented here are only those related to this change.

Thanks a lot for looking into this!!

**Comment by prozessor13** [comment]
How can i do this? And beside this, this branch do not belongs to the *paralell_rtt2* branch, it would also work to the *main* branch. But dont know how to change this either :/

**Comment by HarelM** [comment]
Sure, I'll help with that.

**Comment by HarelM** [comment]
Ok, there are two options here:
1. The one I think is better: Switch to main branch in your repo, then create from it a new branch and make the changes there and open a different pull request - this way the changes are not coupled with the merge of the other fix
2. More complicated: Create a branch here, merge the other fix to this branch and not to main, and change the base branch of this PR to the branch you created in this repo.

Both options are valid, but in I think the first one is simpler and better.

I looked at the second check-in which I think is the relevant change in terms of adding the frame to the terrain and in the glsl code I saw some hard coded 450 which is probably related to the dead sea. While I guess this will work for the current use cases it might not be sufficient for maps that has terrain data for oceans etc. I've seen these in the past and I don't want to limit the solution here to the dead sea and hard coding 450 in the code.
I'm not sure I know how to properly solve this for all use cases but let's at least try :-)

**Comment by HarelM** [comment]
Let me know if I can help in any other way beside code reviewing.

File: src/render/terrain.ts [hunk 1]
```diff
@@ -358,6 +317,27 @@ export default class Terrain {
             indexArray.emplaceBack(x + y, meshSize + x + y + 1, meshSize + x + y + 2);
             indexArray.emplaceBack(x + y, meshSize + x + y + 2, x + y + 1);
         }
+        // add an extra frame around the mesh to avoid stiching on tile boundaries with different zoomlevels
+        // encode z-coordinate into x-coordinate by using x values outside EXTENT
+        // first code-block is for top-bottom frame and second for left-right frame
+        let offset = vertexArray.length, offset2 = offset + (meshSize + 1) * 2;
+        for (let y = 0; y <= 1; y++) for (let x = 0; x <= meshSize; x++) for (let z = 0; z <= 1; z++)
```
Comment (HarelM) [inline src/render/terrain.ts:324]: `for (let y of [0, 1])` is clearer I think, same for `for (let z of [0,1])`.

File: src/render/terrain.ts [hunk 2]
```diff
@@ -358,6 +317,27 @@ export default class Terrain {
             indexArray.emplaceBack(x + y, meshSize + x + y + 1, meshSize + x + y + 2);
             indexArray.emplaceBack(x + y, meshSize + x + y + 2, x + y + 1);
         }
+        // add an extra frame around the mesh to avoid stiching on tile boundaries with different zoomlevels
+        // encode z-coordinate into x-coordinate by using x values outside EXTENT
+        // first code-block is for top-bottom frame and second for left-right frame
+        let offset = vertexArray.length, offset2 = offset + (meshSize + 1) * 2;
```
Comment (HarelM) [inline src/render/terrain.ts:323]: Use separate lines to declare variables.
Can you give better names - not `offset` and `offset2`? It's unclear what they represent.

File: src/render/terrain.ts [hunk 3]
```diff
@@ -358,6 +317,27 @@ export default class Terrain {
...
+        for (let y = 0; y <= 1; y++) for (let x = 0; x <= meshSize; x++) for (let z = 0; z <= 1; z++)
+            vertexArray.emplaceBack(x * delta + z * (EXTENT + 1), y * EXTENT);
+        for (let x = 0; x < meshSize * 2; x += 2) {
+            indexArray.emplaceBack(offset2 + x, offset2 + x + 1, offset2 + x + 3);
+            indexArray.emplaceBack(offset2 + x, offset2 + x + 3, offset2 + x + 2);
+            indexArray.emplaceBack(offset + x, offset + x + 3, offset + x + 1);
+            indexArray.emplaceBack(offset + x, offset + x + 2, offset + x + 3);
+        }
+        offset = vertexArray.length; offset2 = offset + (meshSize + 1) * 2;
+        for (let x = 0; x <= 1; x++) for (let y = 0; y <= meshSize; y++) for (let z = 0; z <= 1; z++)
```
Comment (HarelM) [inline src/render/terrain.ts:333]: These two parts of code seems very similar.
I'm not sure if placing them it in a method with relevant parameters can make it clearer and shorter - it seems that you switch between x and y in these two, not sure this is the only change, but if so, I would consider extracting it to a method with the relevant parameters.

File: src/shaders/terrain.vertex.glsl [hunk 4]
```diff
@@ -6,7 +6,16 @@ varying vec2 v_texture_pos;
...
 
 void main() {
-    v_texture_pos = a_pos / 8192.0; // 8192.0 is the hardcoded vector-tiles coordinates resolution
-    gl_Position = u_matrix * vec4(a_pos, get_elevation(a_pos), 1.0);
+    float extent = 8192.0; // 8192.0 is the hardcoded vector-tiles coordinates resolution
+    vec2 pos = a_pos;
+    float ele = 0.0;
+    if (pos.x > extent) { // check for encoded z values when creating tile frame
+        pos.x -= extent - 1.0;
+        ele = -450.0;
```
Comment (HarelM) [inline src/shaders/terrain.vertex.glsl:14]: As I wrote, I have a problem with this hard coded value.
Is there a way to only add for example 10 (or a bit more) pixels frame from all sides and not go down until the bottom of the dead sea?

**Comment by prozessor13** [comment]
I close this pull-request, because there exists a new one: https://github.com/maplibre/maplibre-gl-js/pull/1666

**Pull request closed without merging**